core/src/package_store.rs

```diff
@@ -0,0 +1,559 @@
+        !self.awaited_packages.is_empty()
+    }
+
+    pub(crate) fn resolve_request(&mut self, request: PackageID, response: Vec<u8>) {
+        if self.awaited_packages.remove(&request) {
+            self.new_packages.insert(request, response);
+        }
+    }
+
+    pub(crate) fn reject_request<E>(&mut self, request: PackageID, response: E)
+    where
+        E: Error + Send + 'static,
+    {
+        if self.awaited_packages.remove(&request) {
+            self.failed_packages
+                .push(CoreError::Resolve(request.name.clone(), Box::new(response)));
+        }
+    }
+}
+
+impl From<&str> for PackageID {
+    fn from(s: &str) -> Self {
+        #[inline]
+        fn prefix<'a, T>(s: &'a str, prefix: &'static str, t: T) -> Option<(&'a str, T)> {
+            s.starts_with(prefix)
+                .then(|| (s.split_at(prefix.len()).1, t))
+        }
+
+        let (name, target) = None
+            .or(prefix(s, "pkg:", PackageSource::Registry))
+            .or(prefix(s, "pkgs:", PackageSource::Registry))
+            .or(prefix(s, "prelude:", PackageSource::Standard))
+            .or(prefix(s, "std:", PackageSource::Standard))
+            .or_else(|| {
+                (s.starts_with("http://") | s.starts_with("https://"))
+                    .then_some((s, PackageSource::Url))
+            })
+            .unwrap_or((s, PackageSource::Local));
+        let name = name.to_string();
+        PackageID {
+            name,
+            source: target,
+        }
+    }
+}
+
+impl From<String> for PackageID {
+    fn from(value: String) -> Self {
+        value.as_str().into()
+    }
+}
+
+impl FromStr for PackageID {
+    type Err = Infallible;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(s.into())
+    }
+}
+
+impl From<&Import> for PackageID {
+    fn from(value: &Import) -> Self {
+        PackageID::from_str(&value.name).unwrap()
+    }
+}
+
+impl From<&Hide> for PackageID {
+    fn from(value: &Hide) -> Self {
+        PackageID::from_str(&value.name).unwrap()
+    }
+}
+
+impl From<Import> for PackageID {
+    fn from(value: Import) -> Self {
+        PackageID::from_str(&value.name).unwrap()
+    }
+}
+
+impl From<Hide> for PackageID {
+    fn from(value: Hide) -> Self {
+        PackageID::from_str(&value.name).unwrap()
+    }
+}
+
+pub trait Resolve {
+    /// The implementor should resolve the given ResolveTasks, and may do so sync or async
+    fn resolve_all(&self, paths: Vec<ResolveTask>);
+}
+
+#[derive(Debug)]
+pub struct ResolveTask {
+    manager: Arc<Mutex<PackageStore>>,
+    pub package: PackageID,
+    resolved: bool,
+}
+
+impl ResolveTask {
+    pub fn complete<E>(self, result: Result<Vec<u8>, E>)
+    where
+        E: Error + Send + 'static,
+    {
+        match result {
+            Ok(result) => self.resolve(result),
+            Err(error) => self.reject(error),
+        }
+    }
+
+    pub fn resolve(mut self, result: Vec<u8>) {
```

Comment:
Instead of using `mem::take` etc could we instead just own self?
```suggestion
    pub fn resolve(self, result: Vec<u8>) {
```

Reply:
Hmm I don't really understand how that would work. We need to either get the package ID or clone it, to get an owned ID to call resolve_request. We somehow thus needs to get the ID from `self`. Is it better to take the reference and clone it? `take` will replace it with default which is pretty cheap.